rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,221 @@
+
+We will continue to use the new development and deployment pipeline created for the trials. Much of the functionality is defined in [a new `Deploy_App_Downstream` Jenkins job](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/modules/govuk_jenkins/templates/jobs/deploy_app_downstream.yaml.erb), which works like this:
+
+1. **Integration**
+
+   1. Deployment runs.
+   1. [**new**] Wait for the Smoke tests for the app to pass [[1](https://github.com/alphagov/govuk-puppet/pull/10387)].
+   1. [**new**] Remotely trigger the same deployment in Staging.
+
+1. **Staging**
+
+   1. [**new**] Repeat as for Integration, but to Production.
+
+We propose to apply this automation across our apps. Given the scale of the change, we need to think carefully about how we can make it consistently safe and secure, before we enable CD for an app.
+
+### Safety concerns
+
+The person who merges a PR must be confident that it is safe to do so. As part of the trial of CD, [we added PR templates to each affected app](https://github.com/alphagov/smart-answers/pull/4758), to explicitly warn developers about the automatic deployment.
+
+However, we also want to ensure that a faulty PR merged in haste can be caught before it impacts the live site. There are a number of checks we can perform to help ensure this, without aiming for perfection.
+
+> **We won't check alerts.** At the time of writing, we have no evidence that checking Icinga alerts has any advantage over the checks discussed below. We can investigate such a check if we find evidence [[1](https://github.com/alphagov/govuk-puppet/pull/10597)].
+
+> **We will only consider _live_ code.** Interactions due to manual and scheduled tasks can be checked manually, using a branch deployment. In the worst case, we should have ample time to fix any faults that develop.
+
+#### Check: Internal repo tests pass (i.e. features are tested in detail)
+
+For the purpose of automating deployments:
+
+- We will use code coverage as our indicator of previous, thorough testing. While an app may have a [rigorous testing strategy](https://github.com/alphagov/content-publisher/blob/69bed3535de1e65bda9d5909a3bc56a319e46576/docs/testing-strategy.md), it's hard to know if this has been applied consistently as the app has evolved.
+
+- We will require a minimum level of coverage that is realistic for all of our apps. We will not pursue this once CD is enabled, as doing so could encourage [a lack of rigour in the way we write our tests](https://jeroenmols.com/blog/2017/11/28/coveragproblem/).
+
+- We will not check coverage of periphery code e.g. CSS, JavaScript. This is because it has a low footprint in our codebases, and it's hard to measure coverage for: the frameworks we use don't support it.
+
+An app has "enough" of these tests when:
+
+- Its code coverage exceeds 95% at the point when CD is enabled.
```

Comment:
One thing that is commonly missed in tests is the actual rendering of the page. The default behaviour in RSpec controller tests for example is not to render the page; it is something that can be [enabled](https://relishapp.com/rspec/rspec-rails/docs/controller-specs/render-views).

As an example, most of the tests of flows in Smart Answers do not render the page. Most of the ruby code is tested, but a miss-configuration in an erb file that would cause a page not to load (misspelled method name for example) may not cause tests to fail.

Therefore, I think it is good to ensure at least once for a page, to test that it does render. Without this an application can have 100% coverage but be unable to render a page!

Reply:
Thanks @reggieb for raising this - I had no idea this was a problem. It turns out this behaviour is unique to Smart Answers: most of our apps have [feature tests](https://github.com/alphagov/content-publisher/tree/master/spec/features) that _do_ render pages and check what's on them.

Since this appears to be a special case, I don't think we need general criteria around it. However, it is alarming that a large portion of the Smart Answers codebase may not be tested. I'm not very familiar with Smart Answers, though.

@theseanything could you help out here? I'd like to understand how much of a problem this is: the spirit of the criteria around coverage is that _most_ of the code in a repo should be covered to enable CD (or leave it enabled). If there is a significant gap in testing, then I think we have to make a special provision for Smart Answers e.g.

- Disable CD while we investigate how we can significantly improve the test coverage for ERB files; or

- Leave CD enabled and document this as an accepted risk (will require input from @bevanloon or @richardTowers).

I think we have the good track record of having CD for Smart Answers on our side, so there's that in favour of leaving things as they are. To be clear: I don't think we should leave CD enabled if we think there is a big gap in testing, as then there is no motivation for us to follow through on fixing that after the RFC concludes.

Reply:
Yup - lack of full integration tests for each smart answer flow (e.g. headless browser completing a flow) is a definite gap in our testing.

Whilst we should have them, I don't think there has been _any significant increase_ in risk of issues making it to production since enabling CD for smart answers.  

If we were to disable CD: 
- It’s still unfeasible to manually check every flow and every possibility. So changes to the core of smart answers (which can effect any flow) carry the same risk of issues making it to production. I.e. we had no way to prevent this before.

- For changes to individual flows our process involves the use of preview apps. Or the draft stack for new flows. The equivalent of checking a flow in integration - and no significant change in risk. 

I can’t see disabling CD reducing this risk and I can’t see why enabling CD should be motivation to get this work done. We should have other ways to prioritise this work (tech debt board?) that aren’t steps backwards - as either way we won't be improving what we really care about (i.e. risk of issues making it production.) Our team is currently investigating whether we can add these missing tests.

Reply:
Thanks for your thoughts @theseanything. 

I don't really want to turn CD off either, especially as Smart Answers seems to be a great example of a whole team adopting it responsibly - with a healthy mixture of manual and automated testing. 

I should still emphasise that enabling CD for Smart Answers was only ever a _trial_. All trials come to an end, and we always said they were just a stepping stone towards an RFC, where we thoroughly consider which apps should have it. An app being in the trial shouldn't mean it gets certified as safe, without further thought.

However, that doesn't mean we can't have special cases, and turning CD off doesn't have to be the outcome. In this case, I think you're saying we should leave it as-is because:

- We're aware of the problem and we have a process in place to manage the risk.
- The risk is unchanged with/out CD, due to the complexity of the code under test.
- Your team is also investigating how we can further improve the situation with tests.

That's good enough for me.

@reggieb does that address your concern?

Reply:
I'm happy with that.

My comment was more to raise awareness that getting 100% coverage did not mean that you had a guarantee that everything was tested. In my experience, not testing pages actually render is common and an excellent example of a condition where coverage (as determined by the likes of *rcov*) could be 100% but still leave large sections of code and functionality untested.

I used Smart Answers as an example, as it's the app I'm currently working on and one where the lack of page rendering in tests has caught me out.